### PR TITLE
feat: 1p secret provider to return the entire entry as JSON. fixes #1498

### LINKED ISSUE #1498: Extend 1Password secret provider to return the entire entry as JSON
Currently a single field must be specified as `op://<vault>/<entry>/<field>`, which is always treated as a single JSON string.

But sometimes we want a composite value. A proposal @mistermoe and I came up with is that if the reference is to the entry itself, eg. `op://<vault>/<entry>`, FTL would retrieve all fields as a JSON object.

eg. Given the following entry as JSON:

```json
{
  "id": "xyz",
  "title": "JSON",
  "tags": ["json"],
  "version": 2,
  "vault": {
    "id": "xyz",
    "name": "Personal"
  },
  "category": "LOGIN",
  "last_edited_by": "xyz",
  "created_at": "2024-05-15T19:12:04Z",
  "updated_at": "2024-05-15T19:19:54Z",
  "additional_information": "test",
  "fields": [
    {
      "id": "password",
      "type": "CONCEALED",
      "purpose": "PASSWORD",
      "label": "password",
      "value": "aSecretPassword",
      "reference": "op://Personal/JSON/password",
      "password_details": {
        "strength": "VERY_GOOD"
      }
    },
    {
      "id": "username",
      "type": "STRING",
      "purpose": "USERNAME",
      "label": "username",
      "value": "test",
      "reference": "op://Personal/JSON/username"
    },
    {
      "id": "notesPlain",
      "type": "STRING",
      "purpose": "NOTES",
      "label": "notesPlain",
      "reference": "op://Personal/JSON/notesPlain"
    }
  ]
}
```

If the secret reference was `op://Personal/JSON/password` we would return the JSON string quoted value `"aSecretPassword"`.

If the secret reference was `op://Personal/JSON` we would retrieve the JSON object `{"password": "aSecretPassword", "username": "test"}`. Note that `notesPlain` would be omitted because it has no value.

Unfortunately we'll need to manually parse the `op://` reference in order to pass the individual values to `op item get --json`, because it doesn't support `op://` references directly. We'll then need to unmarshal the JSON into an intermediate structure in order to retrieve the fields.

### PR DESCRIPTION
**Pull request by gak**

### CODE REVIEW

File: common/configuration/1password_provider.go [hunk 1]
```diff
@@ -25,6 +27,19 @@ func (OnePasswordProvider) Role() Secrets                               { return
 func (o OnePasswordProvider) Key() string                               { return "op" }
 func (o OnePasswordProvider) Delete(ctx context.Context, ref Ref) error { return nil }
 
+// Load returns either a single field if the op:// reference specifies a field, or all fields if not.
+//
+// A single value/password:
+// op://Personal/With Spaces
```
Comment (alecthomas) [inline common/configuration/1password_provider.go:33]: I think the example here and below are swapped

File: common/configuration/1password_provider.go [hunk 2]
```diff
@@ -36,16 +51,46 @@ func (o OnePasswordProvider) Load(ctx context.Context, ref Ref, key *url.URL) ([
...
+		return nil, err
 	}
-	return json.Marshal(string(output))
+
+	// Filter out anything without a value
+	filtered := slices.Filter(full, func(e entry) bool {
+		return e.Value != ""
+	})
+	// Map to id: value
+	var mapped = make(map[string]string)
```
Comment (alecthomas) [inline common/configuration/1password_provider.go:88]: If you know the number of final entries, it's typical to set the capacity with eg. `make(map[string]string, len(filtered))`

File: common/configuration/1password_provider.go [hunk 3]
```diff
@@ -61,3 +106,93 @@ func (o OnePasswordProvider) Store(ctx context.Context, ref Ref, value []byte) (
...
+//
+// Secret references are case-insensitive and support the following characters:
+//
+//	alphanumeric characters (a-z, A-Z, 0-9), -, _, . and the whitespace character
+//
+// If an item or field name includes a / or an unsupported character, use the item
+// or field's unique identifier (ID) instead of its name.
+//
+// See https://developer.1password.com/docs/cli/secrets-reference-syntax/
+type secretRef struct {
```
Comment (alecthomas) [inline common/configuration/1password_provider.go:152]: Once the other change goes in where we completely manage the secrets, we could probably remove this and just implement our own URI scheme that is valid.

**Review by alecthomas — APPROVED** [review]
Naisu!

File: common/configuration/1password_provider.go (hunk 3 above)
Comment (gak) [reply common/configuration/1password_provider.go:152]: Had the same thought.

**Pull request merged**